src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+        /** option to use optimizing mode or not */
+        optimizingMode
+    }
+
+    private static final Logger logger = Logger.getLogger(Z3SmtSolver.class.getName());
+
+    protected final Context ctx;
+    protected com.microsoft.z3.Optimize solver;
+    protected StringBuilder smtFileContents;
+
+    protected static final String z3Program = "z3";
+    protected boolean optimizingMode;
+
+    /** This field indicates that whether we are going to explain unsatisfiable.*/
+    protected boolean getUnsatCore;
+
+    // used in non-optimizing mode to find unsat constraints
+    protected final Map<String, Constraint> serializedConstraints = new HashMap<>();
+
+    /** The set of IDs of unsat core constraints */
+    protected final List<String> unsatConstraintIDs = new ArrayList<>();
+
+    // file is written at projectRootFolder/constraints.smt
+    protected static final String pathToProject = new File("").getAbsolutePath();
+    protected static final String constraintsFile = pathToProject + "/z3Constraints.smt";
+    protected static final String constraintsUnsatCoreFile =
+            pathToProject + "/z3ConstraintsUnsatCore.smt";
+    protected static final String constraintsStatsFile = pathToProject + "/z3ConstraintsGlob.smt";
+
+    // timing statistics variables
+    protected long serializationStart;
+    protected long serializationEnd;
+    protected long solvingStart;
+    protected long solvingEnd;
+
+    /** Whether the timeout arg should be set */
+    protected boolean withTimeout = false;
+
+    public Z3SmtSolver(
+            SolverEnvironment solverEnvironment,
+            Collection<Slot> slots,
+            Collection<Constraint> constraints,
+            Z3SmtFormatTranslator<SlotEncodingT, SlotSolutionT> z3SmtFormatTranslator,
+            Lattice lattice) {
+        super(solverEnvironment, slots, constraints, z3SmtFormatTranslator, lattice);
+
+        Map<String, String> z3Args = new HashMap<>();
+
+        if (withTimeout) {
+            z3Args.put("timeout", Integer.toString(timeout()));
+        }
+        // creating solver
+	    ctx = new Context(z3Args);
+
+        z3SmtFormatTranslator.init(ctx);
+    }
+
+    protected int timeout() {
+        return 2 * 60 * 1000; // timeout of 2 mins by default
+    }
+
+    // Main entry point
+    @Override
+    public Map<Integer, AnnotationMirror> solve() {
+        // serialize based on user choice of running in optimizing or non-optimizing mode
+        optimizingMode = solverEnvironment.getBoolArg(Z3SolverEngineArg.optimizingMode);
+        getUnsatCore = false;
+
+        if (optimizingMode) {
+            logger.fine("Encoding for optimizing mode");
+        } else {
+            logger.fine("Encoding for non-optimizing mode");
+        }
+
+        serializeSMTFileContents();
+
+        solvingStart = System.currentTimeMillis();
+        // in Units, if the status is SAT then there must be output in the model
```

Comment:
This comment doesn't seem helpful. Can you clarify or remove? Maybe it should be moved to the `result == null` below?